File: src/org/thoughtcrime/securesms/recipients/RecipientDetails.java
package org.thoughtcrime.securesms.recipients;

import android.content.Context;
import android.net.Uri;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import org.thoughtcrime.securesms.color.MaterialColor;
import org.thoughtcrime.securesms.database.RecipientDatabase.InsightsBannerTier;
import org.thoughtcrime.securesms.crypto.ProfileKeyUtil;
import org.thoughtcrime.securesms.database.RecipientDatabase.RecipientSettings;
import org.thoughtcrime.securesms.database.RecipientDatabase.RegisteredState;
import org.thoughtcrime.securesms.database.RecipientDatabase.UnidentifiedAccessMode;
import org.thoughtcrime.securesms.database.RecipientDatabase.VibrateState;
import org.thoughtcrime.securesms.util.TextSecurePreferences;
import org.thoughtcrime.securesms.util.Util;
import org.whispersystems.libsignal.util.guava.Optional;

import java.util.LinkedList;
import java.util.List;
import java.util.UUID;

public class RecipientDetails {

  final UUID                   uuid;
  final String                 e164;
  final String                 email;
  final String                 groupId;
  final String                 name;
  final String                 customLabel;
  final Uri                    systemContactPhoto;
  final Uri                    contactUri;
  final Optional<Long>         groupAvatarId;
  final MaterialColor          color;
  final Uri                    messageRingtone;
  final Uri                    callRingtone;
  final long                   mutedUntil;
  final VibrateState           messageVibrateState;
  final VibrateState           callVibrateState;
  final boolean                blocked;
  final int                    expireMessages;
  final List<Recipient>        participants;
  final String                 profileName;
  final Optional<Integer>      defaultSubscriptionId;
  final RegisteredState        registered;
  final byte[]                 profileKey;
  final String                 profileAvatar;
  final boolean                profileSharing;
  final boolean                systemContact;
  final boolean                isLocalNumber;
  final String                 notificationChannel;
  final UnidentifiedAccessMode unidentifiedAccessMode;
  final boolean                forceSmsSelection;
  final boolean                uuidSuported;
  final InsightsBannerTier     insightsBannerTier;

  RecipientDetails(@NonNull Context context,
                   @Nullable String name,
                   @NonNull Optional<Long> groupAvatarId,
                   boolean systemContact,
                   boolean isLocalNumber,
                   @NonNull RecipientSettings settings,
                   @Nullable List<Recipient> participants)
  {
    this.groupAvatarId                   = groupAvatarId;
    this.systemContactPhoto              = Util.uri(settings.getSystemContactPhotoUri());
    this.customLabel                     = settings.getSystemPhoneLabel();
    this.contactUri                      = Util.uri(settings.getSystemContactUri());
    this.uuid                            = settings.getUuid();
    this.e164                            = settings.getE164();
    this.email                           = settings.getEmail();
    this.groupId                         = settings.getGroupId();
    this.color                           = settings.getColor();
    this.messageRingtone                 = settings.getMessageRingtone();
    this.callRingtone                    = settings.getCallRingtone();
    this.mutedUntil                      = settings.getMuteUntil();
    this.messageVibrateState             = settings.getMessageVibrateState();
    this.callVibrateState                = settings.getCallVibrateState();
    this.blocked                         = settings.isBlocked();
    this.expireMessages                  = settings.getExpireMessages();
    this.participants                    = participants == null ? new LinkedList<>() : participants;
    this.profileName                     = isLocalNumber ? TextSecurePreferences.getProfileName(context) : settings.getProfileName();
    this.defaultSubscriptionId           = settings.getDefaultSubscriptionId();
    this.registered                      = settings.getRegistered();
    this.profileKey                      = isLocalNumber ? ProfileKeyUtil.getProfileKey(context) : settings.getProfileKey();
    this.profileAvatar                   = settings.getProfileAvatar();
    this.profileSharing                  = settings.isProfileSharing();
    this.systemContact                   = systemContact;
    this.isLocalNumber                   = isLocalNumber;
    this.notificationChannel             = settings.getNotificationChannel();
    this.unidentifiedAccessMode          = settings.getUnidentifiedAccessMode();
    this.forceSmsSelection               = settings.isForceSmsSelection();
    this.uuidSuported                    = settings.isUuidSupported();
    this.insightsBannerTier              = settings.getInsightsBannerTier();

    if (name == null) this.name = settings.getSystemDisplayName();
    else              this.name = name;
  }

  public RecipientDetails() {
    this.groupAvatarId          = null;
    this.systemContactPhoto     = null;
    this.customLabel            = null;
    this.contactUri             = null;
    this.uuid                   = null;
    this.e164                   = null;
    this.email                  = null;
    this.groupId                = null;
    this.color                  = null;
    this.messageRingtone        = null;
    this.callRingtone           = null;
    this.mutedUntil             = 0;
    this.messageVibrateState    = VibrateState.DEFAULT;
    this.callVibrateState       = VibrateState.DEFAULT;
    this.blocked                = false;
    this.expireMessages         = 0;
    this.participants           = new LinkedList<>();
    this.profileName            = null;
    this.insightsBannerTier     = InsightsBannerTier.TIER_TWO;
    this.defaultSubscriptionId  = Optional.absent();
    this.registered             = RegisteredState.UNKNOWN;
    this.profileKey             = null;
    this.profileAvatar          = null;
    this.profileSharing         = false;
    this.systemContact          = true;
    this.isLocalNumber          = false;
    this.notificationChannel    = null;
    this.unidentifiedAccessMode = UnidentifiedAccessMode.UNKNOWN;
    this.forceSmsSelection      = false;
    this.name                   = null;
    this.uuidSuported           = false;
  }
}

File: app/src/main/java/org/thoughtcrime/securesms/recipients/RecipientDetails.kt
package org.thoughtcrime.securesms.recipients

import android.content.Context
import android.net.Uri
import androidx.annotation.VisibleForTesting
import org.signal.libsignal.zkgroup.profiles.ExpiringProfileKeyCredential
import org.thoughtcrime.securesms.badges.models.Badge
import org.thoughtcrime.securesms.conversation.colors.AvatarColor
import org.thoughtcrime.securesms.conversation.colors.ChatColors
import org.thoughtcrime.securesms.database.RecipientTable.MentionSetting
import org.thoughtcrime.securesms.database.RecipientTable.PhoneNumberSharingState
import org.thoughtcrime.securesms.database.RecipientTable.RegisteredState
import org.thoughtcrime.securesms.database.RecipientTable.UnidentifiedAccessMode
import org.thoughtcrime.securesms.database.RecipientTable.VibrateState
import org.thoughtcrime.securesms.database.model.DistributionListId
import org.thoughtcrime.securesms.database.model.GroupRecord
import org.thoughtcrime.securesms.database.model.ProfileAvatarFileDetails
import org.thoughtcrime.securesms.database.model.RecipientRecord
import org.thoughtcrime.securesms.groups.GroupId
import org.thoughtcrime.securesms.keyvalue.SignalStore
import org.thoughtcrime.securesms.profiles.ProfileName
import org.thoughtcrime.securesms.recipients.Recipient.HiddenState
import org.thoughtcrime.securesms.service.webrtc.links.CallLinkRoomId
import org.thoughtcrime.securesms.util.TextSecurePreferences
import org.thoughtcrime.securesms.util.Util
import org.thoughtcrime.securesms.wallpaper.ChatWallpaper
import org.whispersystems.signalservice.api.push.ServiceId.ACI
import org.whispersystems.signalservice.api.push.ServiceId.PNI
import java.util.LinkedList
import java.util.Optional

class RecipientDetails private constructor(
  @JvmField val aci: ACI?,
  @JvmField val pni: PNI?,
  @JvmField val username: String?,
  @JvmField val e164: String?,
  @JvmField val email: String?,
  @JvmField val groupId: GroupId?,
  @JvmField val distributionListId: DistributionListId?,
  /** Used for groups, dlists, and call links */
  @JvmField val groupName: String?,
  @JvmField val systemContactName: String?,
  @JvmField val customLabel: String?,
  @JvmField val systemContactPhoto: Uri?,
  @JvmField val contactUri: Uri?,
  @JvmField val groupAvatarId: Optional<Long>,
  @JvmField val messageRingtone: Uri?,
  @JvmField val callRingtone: Uri?,
  @JvmField val mutedUntil: Long,
  @JvmField val messageVibrateState: VibrateState,
  @JvmField val callVibrateState: VibrateState,
  @JvmField val blocked: Boolean,
  @JvmField val expireMessages: Int,
  @JvmField val participantIds: List<RecipientId>,
  @JvmField val profileName: ProfileName,
  @JvmField val registered: RegisteredState,
  @JvmField val profileKey: ByteArray?,
  @JvmField val expiringProfileKeyCredential: ExpiringProfileKeyCredential?,
  @JvmField val profileAvatar: String?,
  @JvmField val profileAvatarFileDetails: ProfileAvatarFileDetails,
  @JvmField val profileSharing: Boolean,
  @JvmField val hiddenState: HiddenState,
  @JvmField val isActiveGroup: Boolean,
  @JvmField val lastProfileFetch: Long,
  @JvmField val isSelf: Boolean,
  @JvmField val notificationChannel: String?,
  @JvmField val unidentifiedAccessMode: UnidentifiedAccessMode,
  @JvmField val capabilities: RecipientRecord.Capabilities,
  @JvmField val storageId: ByteArray?,
  @JvmField val mentionSetting: MentionSetting,
  @JvmField val wallpaper: ChatWallpaper?,
  @JvmField val chatColors: ChatColors?,
  @JvmField val avatarColor: AvatarColor,
  @JvmField val about: String?,
  @JvmField val aboutEmoji: String?,
  @JvmField val systemProfileName: ProfileName,
  @JvmField val extras: Optional<Recipient.Extras>,
  @JvmField val hasGroupsInCommon: Boolean,
  @JvmField val badges: List<Badge>,
  @JvmField val isReleaseChannel: Boolean,
  @JvmField val needsPniSignature: Boolean,
  @JvmField val callLinkRoomId: CallLinkRoomId?,
  @JvmField val groupRecord: Optional<GroupRecord>,
  @JvmField val phoneNumberSharing: PhoneNumberSharingState
) {

  @VisibleForTesting
  constructor(
    groupName: String?,
    systemContactName: String?,
    isSelf: Boolean,
    registeredState: RegisteredState,
    record: RecipientRecord,
    participantIds: List<RecipientId>?,
    isReleaseChannel: Boolean,
    avatarColor: AvatarColor?,
    groupRecord: Optional<GroupRecord>
  ) : this(
    groupAvatarId = groupRecord.map { if (it.hasAvatar()) it.avatarId else null },
    systemContactPhoto = Util.uri(record.systemContactPhotoUri),
    customLabel = record.systemPhoneLabel,
    contactUri = Util.uri(record.systemContactUri),
    aci = record.aci,
    pni = record.pni,
    username = record.username,
    e164 = record.e164,
    email = record.email,
    groupId = record.groupId,
    distributionListId = record.distributionListId,
    messageRingtone = record.messageRingtone,
    callRingtone = record.callRingtone,
    mutedUntil = record.muteUntil,
    messageVibrateState = record.messageVibrateState,
    callVibrateState = record.callVibrateState,
    blocked = record.isBlocked,
    expireMessages = record.expireMessages,
    participantIds = participantIds ?: LinkedList(),
    isActiveGroup = groupRecord.map { it.isActive }.orElse(false),
    profileName = record.signalProfileName,
    registered = registeredState,
    profileKey = record.profileKey,
    expiringProfileKeyCredential = record.expiringProfileKeyCredential,
    profileAvatar = record.signalProfileAvatar,
    profileAvatarFileDetails = record.profileAvatarFileDetails,
    profileSharing = record.profileSharing,
    hiddenState = record.hiddenState,
    lastProfileFetch = record.lastProfileFetch,
    isSelf = isSelf,
    notificationChannel = record.notificationChannel,
    unidentifiedAccessMode = record.unidentifiedAccessMode,
    capabilities = record.capabilities,
    storageId = record.storageId,
    mentionSetting = record.mentionSetting,
    wallpaper = record.wallpaper,
    chatColors = record.chatColors,
    avatarColor = avatarColor ?: record.avatarColor,
    about = record.about,
    aboutEmoji = record.aboutEmoji,
    systemProfileName = record.systemProfileName,
    groupName = groupName,
    systemContactName = systemContactName,
    extras = Optional.ofNullable(record.extras),
    hasGroupsInCommon = record.hasGroupsInCommon,
    badges = record.badges,
    isReleaseChannel = isReleaseChannel,
    needsPniSignature = record.needsPniSignature,
    callLinkRoomId = record.callLinkRoomId,
    groupRecord = groupRecord,
    phoneNumberSharing = record.phoneNumberSharing
  )

  companion object {
    @JvmStatic
    fun forIndividual(context: Context, record: RecipientRecord): RecipientDetails {
      val isSelf = record.e164 != null && record.e164 == SignalStore.account().e164 || record.aci != null && record.aci == SignalStore.account().aci
      val isReleaseChannel = record.id == SignalStore.releaseChannelValues().releaseChannelRecipientId
      var registeredState = record.registered

      if (isSelf) {
        registeredState = if (SignalStore.account().isRegistered && !TextSecurePreferences.isUnauthorizedReceived(context)) {
          RegisteredState.REGISTERED
        } else {
          RegisteredState.NOT_REGISTERED
        }
      }

      return RecipientDetails(
        groupName = null,
        systemContactName = record.systemDisplayName,
        isSelf = isSelf,
        registeredState = registeredState,
        record = record,
        participantIds = null,
        isReleaseChannel = isReleaseChannel,
        avatarColor = null,
        groupRecord = Optional.empty()
      )
    }

    @JvmStatic
    fun forGroup(groupRecord: GroupRecord, recipientRecord: RecipientRecord): RecipientDetails {
      return RecipientDetails(
        groupName = groupRecord.title,
        systemContactName = null,
        isSelf = false,
        registeredState = recipientRecord.registered,
        record = recipientRecord,
        participantIds = groupRecord.members,
        isReleaseChannel = false,
        avatarColor = null,
        groupRecord = Optional.of(groupRecord)
      )
    }

    @JvmStatic
    fun forDistributionList(title: String?, members: List<RecipientId>?, record: RecipientRecord): RecipientDetails {
      return RecipientDetails(
        groupName = title,
        systemContactName = null,
        isSelf = false,
        registeredState = record.registered,
        record = record,
        participantIds = members,
        isReleaseChannel = false,
        avatarColor = null,
        groupRecord = Optional.empty()
      )
    }

    @JvmStatic
    fun forCallLink(name: String?, record: RecipientRecord, avatarColor: AvatarColor): RecipientDetails {
      return RecipientDetails(
        groupName = name,
        systemContactName = null,
        isSelf = false,
        registeredState = record.registered,
        record = record,
        participantIds = emptyList(),
        isReleaseChannel = false,
        avatarColor = avatarColor,
        groupRecord = Optional.empty()
      )
    }

    @JvmStatic
    fun forUnknown(): RecipientDetails {
      return RecipientDetails(
        groupAvatarId = Optional.empty(),
        systemContactPhoto = null,
        customLabel = null,
        contactUri = null,
        aci = null,
        pni = null,
        username = null,
        e164 = null,
        email = null,
        groupId = null,
        distributionListId = null,
        messageRingtone = null,
        callRingtone = null,
        mutedUntil = 0,
        messageVibrateState = VibrateState.DEFAULT,
        callVibrateState = VibrateState.DEFAULT,
        blocked = false,
        expireMessages = 0,
        participantIds = LinkedList(),
        profileName = ProfileName.EMPTY,
        registered = RegisteredState.UNKNOWN,
        profileKey = null,
        expiringProfileKeyCredential = null,
        profileAvatar = null,
        profileAvatarFileDetails = ProfileAvatarFileDetails.NO_DETAILS,
        profileSharing = false,
        hiddenState = HiddenState.NOT_HIDDEN,
        lastProfileFetch = 0,
        isSelf = false,
        notificationChannel = null,
        unidentifiedAccessMode = UnidentifiedAccessMode.UNKNOWN,
        groupName = null,
        capabilities = RecipientRecord.Capabilities.UNKNOWN,
        storageId = null,
        mentionSetting = MentionSetting.ALWAYS_NOTIFY,
        wallpaper = null,
        chatColors = null,
        avatarColor = AvatarColor.UNKNOWN,
        about = null,
        aboutEmoji = null,
        systemProfileName = ProfileName.EMPTY,
        systemContactName = null,
        extras = Optional.empty(),
        hasGroupsInCommon = false,
        badges = emptyList(),
        isReleaseChannel = false,
        needsPniSignature = false,
        isActiveGroup = false,
        callLinkRoomId = null,
        groupRecord = Optional.empty(),
        phoneNumberSharing = PhoneNumberSharingState.UNKNOWN
      )
    }
  }
}
